### PR TITLE
feat: support catalog source

### PR DESCRIPTION
**Pull request by y-lakhdar**
_Big PR, but 68% unit tests_

## Proposed changes
Adding support for the catalog source.
Since documents can be pushed to the catalog source via the standard file container flow and the stream API, I created a second strategy `StreamChunkStrategy`.

The idea here is to use the Strategy pattern to prevent code duplication but also because the relation between the source and the upload strategy is not necessarily one-to-one. You can see the high level diagram [here](https://docs.google.com/drawings/d/14zlQUuJObHUqdMu4Bx92Ht_8bbcIdfAvSg0CCBCgstY/edit).

Eventually, the catalog and/or source may leverage other strategies to update documents.

## Testing
* [x] Manual tests
* [x] Unit tests

### CODE REVIEW

File: src/source/documentUploader.ts [hunk 1]
```diff
@@ -10,7 +11,7 @@ export async function uploadBatch(
   strategy: UploadStrategy,
   batch: BatchUpdateDocuments,
   createFields = true
-) {
+): Promise<AxiosResponse> {
   if (createFields) {
     const analyser = new FieldAnalyser(platformClient);
```
Comment (louis-bompart) [inline src/source/documentUploader.ts:16]: I just noticed, tho reallly minor.
Which 'flavoor' of English is preferred for public APIs?
British (so -ise) or American (-ize)?

File: src/uploadStrategy/fileContainerStrategy.ts [hunk 2]
```diff
@@ -21,12 +21,14 @@ export class FileContainerStrategy implements UploadStrategy {
...
   ) {}
 
-  public async uploadBatch(batch: BatchUpdateDocuments) {
+  public async upload(batch: BatchUpdateDocuments) {
     const fileContainer = await this.createFileContainer();
     await uploadContentToFileContainer(fileContainer, batch);
     return this.pushFileContainerContent(fileContainer);
   }
 
+  public async postUpload() {}
```
Comment (louis-bompart) [inline src/uploadStrategy/fileContainerStrategy.ts:30]: ? if empty, then abstract?

File: src/uploadStrategy/strategy.ts [hunk 3]
```diff
@@ -7,5 +7,12 @@ export interface UploadStrategy {
...
    */
-  uploadBatch: (batch: BatchUpdateDocuments) => Promise<AxiosResponse>;
+  upload: (batch: BatchUpdateDocuments) => Promise<AxiosResponse>;
+
+  /**
+   * Async operation to run once the batch upload is complete
+   *
+   * @memberof UploadStrategy
+   */
+  postUpload: () => Promise<void>;
```
Comment (louis-bompart) [inline src/uploadStrategy/strategy.ts:17]: I see that `FileContainerStrategy` does virtually not use it.
If that's the case, I would define this as an optional, and the caller of this should check if there's a postUpload before calling it.

File: src/source/batchUploadDocumentsFromFile.ts [hunk 4]
```diff
@@ -54,6 +54,7 @@ export class BatchUploadDocumentsFromFiles {
         await create(platformClient, fields);
       }
       await this.consumer.consume(files);
+      await strategy.postUpload();
```
Comment (louis-bompart) [inline src/source/batchUploadDocumentsFromFile.ts:57]: in relation to https://github.com/coveo/push-api-client.js/pull/117/files#r838077559,

```suggestion
      await strategy.postUpload?.();
```

File: src/source/documentUploader.ts [hunk 5]
```diff
@@ -22,5 +23,8 @@ export async function uploadBatch(
 
     await create(platformClient, fields);
   }
-  return strategy.uploadBatch(batch);
+  const res = await strategy.upload(batch);
+  await strategy.postUpload();
```
Comment (louis-bompart) [inline src/source/documentUploader.ts:27]: in relation to https://github.com/coveo/push-api-client.js/pull/117/files#r838077559,
```suggestion
  await strategy.postUpload?.();
```

**Review by louis-bompart — APPROVED** [review]

File: src/source/documentUploader.ts (hunk 1 above)
Comment (y-lakhdar) [reply src/source/documentUploader.ts:16]: British or course 
image

**Pull request merged**